src/GNDStk/Node.hpp

```diff
@@ -0,0 +1,187 @@
+// -----------------------------------------------------------------------------
+
+template<
+   template<class...> class METADATA_CONTAINER, // container type for metadata
+   template<class...> class CHILDREN_CONTAINER  // container type for children
+>
+class Node {
+   using metaPair = std::pair<std::string,std::string>;
+   using childPtr = std::unique_ptr<Node>;
+
+public:
+
+   // ------------------------
+   // Data
+   // ------------------------
+
+   // Simple node for our tree structure:
+   //    name
+   //    metadata (container of string pairs)
+   //    children (container of pointers to other Node<>s)
+   std::string name;
+   METADATA_CONTAINER<metaPair,std::allocator<metaPair>> metadata;
+   CHILDREN_CONTAINER<childPtr,std::allocator<childPtr>> children;
+
+   // ------------------------
+   // Simple functions
+   // ------------------------
+
+   // clear()
+   Node &clear()
+   {
+      name = "";
+      metadata.clear();
+      children.clear();
+      return *this;
+   }
+
+   // is this node empty?
+   bool empty() const
+   {
+      return name == "" && metadata.size() == 0 && children.size() == 0;
+   }
+
+   // is this a leaf node?
+   bool leaf() const
+   {
+      return children.size() == 0;
+   }
+
+   // copy
+   Node copy() const
+   {
+      Node ret;
+      detail::node2Node(*this,ret);
+      return ret;
+   }
+
+   // normalize
+   Node &normalize();
+
+   // ------------------------
+   // General functions
+   // ------------------------
+
+   // ctor: default
+   Node() { }
+
+   // ctor: move, DELETED copy
+   Node(Node &&) = default;
+   Node(const Node &) = delete;
+
+   // assignment: move, DELETED copy
+   Node &operator=(Node &&) = default;
+   Node &operator=(const Node &) = delete;
+
+   #include "GNDStk/Node/src/add.hpp"
+   #include "GNDStk/Node/src/write.hpp"
+   #include "GNDStk/Node/src/meta.hpp"
+   #include "GNDStk/Node/src/child.hpp"
+   #include "GNDStk/Node/src/call.hpp"
+
+}; // class Node
+
+
+
+// -----------------------------------------------------------------------------
+// operator<<
+// -----------------------------------------------------------------------------
+
+template<
+   template<class...> class METADATA_CONTAINER,
+   template<class...> class CHILDREN_CONTAINER
+>
+inline std::ostream &operator<<(
+   std::ostream &os,
+   const Node<METADATA_CONTAINER,CHILDREN_CONTAINER> &obj
+) {
+   return obj.write(os);
+}
+
+
+
+// -----------------------------------------------------------------------------
+// normalize
+// -----------------------------------------------------------------------------
+
+template<
+   template<class...> class METADATA_CONTAINER,
+   template<class...> class CHILDREN_CONTAINER
+>
+Node<METADATA_CONTAINER,CHILDREN_CONTAINER> &
+Node<METADATA_CONTAINER,CHILDREN_CONTAINER>::normalize()
+{
+   // name
+   detail::strip(name);
+
+   // children
+   auto iter = children.end();
+   for (auto c = children.begin();  c != children.end();  ++c)
```

Comment:
Normally I would, and you'll notice that I do use a range-based for() near the end of the same function, when I visit metadata.

While I'll need to review all this JSON material again to recall precisely why I did things the way I did, it looks like I'm making some decisions and doing some error checking, inside the loop, on the basis of what value the iterator has. I also have a conditional that uses one of the iterators just after the loop. (Which would also be why I'd set up that iterator just before the for(), not inside the for()).

So, while I'm too tired (Friday night) right now to recall the "why," I know I had a reason for doing that. As I've mentioned elsewhere, I'm going to revisit and improve all the JSON material soon, and that especially includes renaming and appropriately splitting normalize() (where this for() appears) into a couple of different functions. I may notice a way to rework the formulation so that a shorter, range-based for() is viable.